sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
+          tableToExtract = JSON_FACTORY.fromString(jsonQueryTempTable, TableReference.class);
+          executeQuery(queryJobId, query, tableToExtract, flattenResults, jobService);
+        } else {
+          tableToExtract = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+        }
+        String extractJobId = jobIdToken + "-extract";
+        List<String> tempFiles = executeExtract(
+            extractJobId, tableToExtract, options.getTempLocation(), jobService);
+
+        TableSchema tableSchema = bqServices.getTableService(bqOptions).getTable(
+            tableToExtract.getProjectId(),
+            tableToExtract.getDatasetId(),
+            tableToExtract.getTableId()).getSchema();
+
+        return createSources(tempFiles, tableSchema);
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        // No way to estimate bytes, so returning 0.
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return false;
+      }
+
+      @Override
+      public BoundedReader<TableRow> createReader(PipelineOptions options) throws IOException {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        if (!Strings.isNullOrEmpty(query)) {
+          return new BigQueryReader(this, bqServices.getReaderFromQuery(
+              bqOptions, query, bqOptions.getProject(), flattenResults));
+        } else {
+          TableReference tableRef = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+          return new BigQueryReader(this, bqServices.getReaderFromTable(bqOptions, tableRef));
+        }
+      }
+
+      @Override
+      public void validate() {
+        // Do nothing, validation is done in BigQuery.Read.
+      }
+
+      @Override
+      public Coder<TableRow> getDefaultOutputCoder() {
+        return TableRowJsonCoder.of();
+      }
+
+      private static void executeQuery(
+          String jobId,
+          String query,
+          TableReference destinationTable,
+          boolean flattenResults,
+          JobService jobService) throws IOException, InterruptedException {
+        JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+        queryConfig
+            .setQuery(query)
+            .setAllowLargeResults(true)
+            .setCreateDisposition("CREATE_IF_NEEDED")
+            .setDestinationTable(destinationTable)
+            .setFlattenResults(flattenResults)
+            .setPriority("BATCH")
+            .setWriteDisposition("WRITE_EMPTY");
+        jobService.startQueryJob(jobId, queryConfig);
+        Job job = jobService.pollJobStatus(
+            destinationTable.getProjectId(), jobId, MAX_JOB_STATUS_POLL_RETRIES);
+        if (parseStatus(job) != Status.SUCCEEDED) {
+          throw new IOException("Query job failed: " + jobId);
+        }
+        return;
+      }
```

Comment:
what is the length of time this is allowed to run with the various maxima? We should be at least as long as BigQuery's limits, if they exist. (A quick search of their docs did not show anything, but I was under the impression it is at least several hours.)